Tella/Components/TopSheetView/TopSheetView.swift

```diff
@@ -0,0 +1,130 @@
+        Group {
+            Spacer()
+                .edgesIgnoringSafeArea(.all)
+                .frame(width: geometry.size.width, height: geometry.size.height)
+                .background(
+                    Color.black.opacity(0.6 * fraction_progress(lowerLimit: 0,
+                                                                upperLimit: kmaxDrag,
+                                                                current: self.currentHeight + currentDrag,
+                                                                inverted: true)
+                    )
+                )
+        }
+    }
+    
+    private var contentView : some View {
+        VStack(alignment: .center) {
+            Spacer()
+                .frame(height: safeArea.top)
+
+            self.content
+            
+            Image("encryption.up")
+                .padding(EdgeInsets(top: 32, leading: 0, bottom: 16, trailing: 0))
+            
+        } .frame(height:self.currentHeight + currentDrag)
+            .frame(maxWidth: .infinity)
+            .background(Styles.Colors.backgroundTab.cornerRadius(30))
+            .clipped()
+            .offset(y: offset)
+            .if(offset == -kMinHeight, transform: { view in
+                view.animation(.interpolatingSpring(stiffness: 1000, damping: 100, initialVelocity: 50))
+            })
+        
+    }
+    private var dragGesture: some Gesture {
+        
+        DragGesture()
+        
+            .onChanged { value in
+                let drag = (value.location.y - value.startLocation.y)
+                
+                if (drag < 0 && self.currentHeight == kMinHeight) {
+                    self.offset = drag
+                } else {
+                    self.currentDrag = drag
+                }
+            }
+            .onEnded({ value in
+                
+                let velocity = (value.velocity)
+                let drag = (value.location.y - value.startLocation.y)
+
+                let newHeight = self.currentHeight + (value.location.y - value.startLocation.y)
+                
+                if (drag < 0 && newHeight < self.kMinHeight && velocity.height < kMinVelocity) {
+                    offset = -kMinHeight
+                    self.presentationMode.wrappedValue.dismiss()
+                }
+                
+                if (drag < 0 && newHeight < self.kMinHeight && velocity.height > kMinVelocity) {
+                    self.offset = 0
+                }
+                
+                if (drag < 0 && newHeight > self.kMinHeight) {
+                    self.currentHeight = kMinHeight
+                }
+                
+                if drag > 0 {
+                    self.currentHeight = kmaxDrag
+                }
+                
+                self.currentDrag = .zero
+            })
+    }
+    
+    private func fraction_progress(lowerLimit: Double = 0, upperLimit:Double, current:Double, inverted:Bool = false) -> Double{
+        var val:Double = 0
```

Comment:
what do you think of using a clamp function for getting the fraction of the value?

```
let clampedCurrent = min(max(current, lowerLimit), upperLimit)
let fraction = (clampedCurrent - lowerLimit) / (upperLimit - lowerLimit)
```

Here we are clamping the value to be within the lower and upper limit and then calculating the fraction using the clamp value.

then we could just return this:


```
return inverted ? fraction : 1 - fraction
```